pkg/virt-handler/vm.go

```diff
@@ -1738,6 +1738,22 @@ func (d *VirtualMachineController) handleMigrationProxy(vmi *v1.VirtualMachineIn
 func (d *VirtualMachineController) processVmUpdate(origVMI *v1.VirtualMachineInstance) error {
 	vmi := origVMI.DeepCopy()
 
+	if vmi.Spec.Domain.CPU != nil {
+		if vmi.Spec.Domain.CPU.Model != "" &&
+			vmi.Spec.Domain.CPU.Model != "host-passthrough" &&
+			vmi.Spec.Domain.CPU.Model != "host-model" &&
+			!d.nodeLabeller.IsCPUModelSupported(vmi.Spec.Domain.CPU.Model) {
+
+			return goerror.New(fmt.Sprintf("vmi requested unsupported cpu model " + vmi.Spec.Domain.CPU.Model))
```

Comment:
You can get to this state pretty easily. Just create label in this format feature.node.kubernetes.io/cpu-model- and put there some model and label node with it. This check checks if the model is really supported. Secondly we need to check if model is supported due to this https://github.com/kubevirt/kubevirt/pull/3403/files#diff-caee6d384498fde52fde15891ac547aaR997 This is according to Jiri Denemark safe only when it is used with cpu models which are reported as usable.

Reply:
Yes,l. I would say that "flaky" is in this case also not fully accurate. There is the mentioned bug https://bugzilla.redhat.com/show_bug.cgi?id=1834714 involved which we have to work around.

Reply:
I understand that there is a specific problem with `qemu64` cpu model, however, I don't really understand why is it being set.
Is it the most suitable one?
According to this test [1] we are not setting the most suitable one, but picking one (any one) from a list. 
afaiu, this is what started this PR. And the test is flaky because these labels are probably formed in a different order each time.

I think the nodes should report the most suitable model and a list of supported. Same way as libvirt does in domcapabilities

```
suitable:
<model fallback='forbid'>Skylake-Client-IBRS</model>
supported:
<mode name='custom' supported='yes'>
      <model usable='yes'>qemu64</model>
.....
.....
```

[1] https://github.com/kubevirt/kubevirt/blob/master/tests/vmi_lifecycle_test.go#L869

Reply:
qemu64 is not the only cpu model with problems, e.g. opteron_g2 has the same issue. And there might be more models about which we don't know (I don't have HW to test it). And the same issue applies to cpu features. 
Labeller original purpose is to show user all supported cpus on the node. User then can select any supported cpus, not onlz the most suitable. And kubevirt should be able to schedule it to correct node, which supports that cpu.

Reply:
> Labeller original purpose is to show user all supported cpus on the node.

I don't think that this is the intent. Think about when you have 60k nodes in a cluster, it's a bit hard to look at one random node and decide on which cpu model to use, especially if there is no guarantee that the VM will land on this random node. And if one does look at this he should know what is the best-suited model for this node.

Yes, these labels are for scheduling, but instead of fixing a test and excluding certain faulty cpu models we are introducing all of this logic which I doubt is required.

Reply:
@vladikr what solution would you propose?

Reply:
> Yes, these labels are for scheduling, but instead of fixing a test and excluding certain faulty cpu models we are introducing all of this logic which I doubt is required.

They are not faulty. They are reported correctly. Libvirt just has a bug when it checks the cpus before startup. This is a workaround until libvirt is fixed. Due to the faulty startup check, all amd models which require smp can't be started on intel cpus, although they are usable.

Reply:
> re is no guarantee that the VM will land on this random node
This is not about looking to random node. This is about using cpu model which user needs (due to e.g. some special features) and ability to schedule vm on the node which can emulate that cpu model.
And as Roman said, libvirt has a bug, which prevents vm to run with certain cpu model eventhough itis supported. And this code allows to vm to run with this model.

Reply:
So, let's do another PR where we just change the CPU model and let's give us a little bit more time to discuss the changes here. Since we don't have an immediate agreement, let's ensure that the test is green again and then improve. @ksimon1 sounds good?

Reply:
https://github.com/kubevirt/kubevirt/pull/3434